src/modules/sdbus/test/message.c

```diff
@@ -0,0 +1,472 @@
+    json_decref (o2);
+    sd_bus_message_unref (m);
+    json_decref (o);
+}
+
+void test_variant_as (sd_bus *bus)
+{
+    sd_bus_message *m;
+    json_t *in;
+    json_t *out;
+    const char *fmt = "v";
+    int rc;
+
+    if (!(in = json_pack ("[[s[sss]]]", "as", "foo", "bar", "baz")))
+        BAIL_OUT ("could not create json object");
+    if (sd_bus_message_new (bus, &m, SD_BUS_MESSAGE_METHOD_RETURN) < 0)
+        BAIL_OUT ("could not create message");
+    ok (sdmsg_write (m, fmt, in) == 0,
+        "sdmsg_write of variant string array works");
+
+    if (sd_bus_message_seal (m, 42, 0) < 0
+        || sd_bus_message_rewind (m, true) < 0)
+        BAIL_OUT ("could not finalize message");
+    diagmsg (m);
+    msgtype_is (m, fmt);
+
+    if (!(out = json_array ()))
+        BAIL_OUT ("could not create json array");
+
+    rc = sdmsg_read (m, fmt, out);
+    diag ("sdmsg_read returned %d", rc);
+    ok (rc == 1,
+        "sdmsg_read works on message containing string array variant");
+
+    diagjson (out);
+
+    const char *type;
+    json_t *val;
+    ok (json_unpack (out, "[[so]]", &type, &val) == 0
+        && streq (type, "as")
+        && json_is_array (val)
+        && json_array_size (val) == 3
+        && json_is_string (json_array_get (val, 0))
+        && json_is_string (json_array_get (val, 1))
+        && json_is_string (json_array_get (val, 2)),
+        "string array variant was translated properly");
+
+    json_decref (out);
+    json_decref (in);
+    sd_bus_message_unref (m);
+}
+
+void test_variant_unknown (sd_bus *bus)
+{
+    sd_bus_message *m;
+    json_t *o;
+    const char *fmt = "svs";
+    uint8_t y[2] = { 99, 100 };
+    int rc;
+
+    if (sd_bus_message_new (bus, &m, SD_BUS_MESSAGE_METHOD_CALL) < 0
+        || sd_bus_message_append (m, "s", "eek") < 0
+        || sd_bus_message_open_container (m, 'v', "a(yy)") < 0
+        || sd_bus_message_open_container (m, 'a', "(yy)") < 0
+        || sd_bus_message_open_container (m, 'r', "yy") < 0
+        || sd_bus_message_append (m, "yy", y[0], y[1]) < 0
+        || sd_bus_message_close_container (m) < 0
+        || sd_bus_message_close_container (m) < 0
+        || sd_bus_message_close_container (m) < 0
+        || sd_bus_message_append (m, "s", "ook") < 0
+        || sd_bus_message_seal (m, 42, 0) < 0
+        || sd_bus_message_rewind (m, true) < 0)
+        BAIL_OUT ("could not create message containing complex variant");
+    diagmsg (m);
+    msgtype_is (m, fmt);
+    if (!(o = json_array ()))
+        BAIL_OUT ("could not create json array");
+
+    rc = sdmsg_read (m, fmt, o);
+    diag ("sdmsg_read returned %d", rc);
+    ok (rc == 1,
+        "sdmsg_read works on message containing complex variant");
+
+    diagjson (o);
+
+    const char *s1, *s2, *type;
+    ok (json_unpack (o, "[s[sn]s]", &s1, &type, &s2) == 0
+        && streq (s1, "eek")
+        && streq (type, "a(yy)")
+        && streq (s2, "ook"),
+        "complex variant was translated to json null");
```

Comment:
took me awhile to grok, but I think you're basically testing if the format is one that sdmsg/interface doesn't know how to handle (b/c you hard code selected ones), it gets converted into json null?

Reply:
Correct.  Maybe a comment to that effect would be a good thing for me to add.